src/ripple/app/tx/impl/CreateTicket.cpp

```diff
@@ -23,105 +23,135 @@
 #include <ripple/protocol/Feature.h>
 #include <ripple/protocol/Indexes.h>
 #include <ripple/protocol/TxFlags.h>
+#include <ripple/protocol/st.h>
 
 namespace ripple {
 
 NotTEC
 CreateTicket::preflight (PreflightContext const& ctx)
 {
-    if (! ctx.rules.enabled(featureTickets))
+    if (! ctx.rules.enabled (featureTicketBatch))
         return temDISABLED;
 
     if (ctx.tx.getFlags() & tfUniversalMask)
         return temINVALID_FLAG;
 
-    auto const ret = preflight1 (ctx);
-    if (!isTesSuccess (ret))
-        return ret;
-
-    if (ctx.tx.isFieldPresent (sfExpiration))
     {
-        if (ctx.tx.getFieldU32 (sfExpiration) == 0)
-        {
-            JLOG(ctx.j.warn()) <<
-                "Malformed transaction: bad expiration";
-            return temBAD_EXPIRATION;
-        }
+        // In C++17: put "count" into if's init-statement, lose outer braces.
+        std::uint32_t const count {ctx.tx[sfCount]};
+        if (count < minValidCount || count > maxValidCount)
+            return temINVALID_COUNT;
+    }
+    {
+        // In C++17: put "ret" into if's init-statement, lose outer braces.
+        NotTEC const ret {preflight1 (ctx)};
+        if (!isTesSuccess (ret))
+            return ret;
     }
-
     return preflight2 (ctx);
 }
 
+TER
+CreateTicket::preclaim(PreclaimContext const& ctx)
+{
+    auto const id = ctx.tx[sfAccount];
+    auto const sleAccountRoot = ctx.view.read(keylet::account(id));
+    if (! sleAccountRoot)
+        return terNO_ACCOUNT;
+
+    // Make sure the TicketCreate would not cause the account to own
+    // too many tickets.
+    auto const optTicketCount {(*sleAccountRoot)[~sfTicketCount]};
+    std::uint32_t const curTicketCount {optTicketCount ? *optTicketCount : 0u};
+    std::uint32_t const addedTickets {ctx.tx[sfCount]};
+
+    if (curTicketCount + addedTickets > maxTicketThreshold)
+        return tecDIR_FULL;
+
+    return tesSUCCESS;
+}
+
 TER
 CreateTicket::doApply ()
 {
-    auto const sle = view().peek(keylet::account(account_));
-    if (! sle)
+    SLE::pointer const sleAccountRoot {
+        view().peek (keylet::account (account_))};
+    if (! sleAccountRoot)    
         return tefINTERNAL;
 
-    // A ticket counts against the reserve of the issuing account, but we
+    // Each ticket counts against the reserve of the issuing account, but we
     // check the starting balance because we want to allow dipping into the
     // reserve to pay fees.
+    std::uint32_t const ticketCount {ctx_.tx[sfCount]};
     {
-        auto const reserve = view().fees().accountReserve(
-            sle->getFieldU32(sfOwnerCount) + 1);
+        XRPAmount const reserve {view().fees().accountReserve (
+            sleAccountRoot->getFieldU32(sfOwnerCount) + ticketCount)};
```

Comment:
This is a minor complaint:

I'm not fond of the use of explicit (direct) initialization when implicit (copy) initialization would get the job done.  My rationale for this preference is that authors of types (such as `XRPAmount`) make the less-safe constructors/conversions explicit while making the more-safe ones implicit.  So users should prefer the more-safe ones, reserving the explicit syntax just for those initializations where one is acknowledging that one has to use a less-safe bit of code.  For example:
```
         XRPAmount const reserve = view().fees().accountReserve (
            sleAccountRoot->getFieldU32(sfOwnerCount) + ticketCount);
```
I know that rippled does not currently follow this guideline at all, and indeed, tends towards the opposite.  And this is my first stab at reversing that trend:  by calling out places where we're using the safe syntax, and moving towards the unsafe syntax.  Note that I do not have an issue here with the replacement of `auto` with `XRPAmount`.

Short (7min) presentation on the potential dangers of overuse of the explicit syntax:  https://www.youtube.com/watch?v=hobFOAehwio

Reply:
Fair, and thanks for pointing this out.  I've been following Nico Josuttis's advice on this, but you make great points.  I'm happy to make the change.

Not to add to your load, but if you could point out all of the places you'd like to see fixed then I can fix them all in a single `[FOLD]` commit, or at least one `[FOLD]` per , um, non-FOLD commit.  Is that a fair request?